### PR TITLE
flatten toc (folder) inclusion dependency map

### PR DESCRIPTION
**Pull request by OsmondJiang**
fix: https://dev.azure.com/ceapex/Engineering/_workitems/edit/96108

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4754)

### CODE REVIEW

File: docs/specs/toc.yml [hunk 1]
```diff
@@ -486,6 +486,41 @@ outputs:
...
+                {
+                    "source": "docs/f1/b.md",
+                    "type": "link"
+                },
+                {
+                    "source": "docs/f1/TOC.md",
+                    "type": "tocInclusion"
+                },
+                {
+                    "source": "docs/f2/a.md",
```
Comment (superyyrrzz) [inline docs/specs/toc.yml]: Not all the dependency in referenced TOC is transit. Actually `toc.md` does not depend on `f1/a.md`.

Comment (OsmondJiang) [reply docs/specs/toc.yml]: good point, seems we need know it's referencing a folder or referencing a toc file

**Comment by OsmondJiang** [comment]
The toc dependencies changes are expected.

image

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 2]
```diff
@@ -419,6 +426,7 @@ private static TocHrefType GetHrefType(string href)
 
         private enum TocHrefType
         {
+            None,
```
Comment (superyyrrzz) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: no use?

File: src/docfx/build/dependency/DependencyType.cs [hunk 3]
```diff
@@ -10,6 +10,7 @@ public enum DependencyType
         UidInclusion, // uid reference with display property
         Inclusion, // token or codesnippet
         Overwrite, // overwrite markdown reference
-        TocInclusion, // toc reference toc
+        TocInclusion, // toc includes toc
+        TocFolderInclusion, // toc includes toc first item
```
Comment (superyyrrzz) [inline src/docfx/build/dependency/DependencyType.cs]: naming: what about `TocFolderReference`? Just to make it clear that it is not an inclusion.

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 4]
```diff
@@ -244,36 +280,35 @@ SourceInfo<string> GetTopicHref(TableOfContentsItem tocInputModel)
                 return default;
             }
 
-            (SourceInfo<string> resolvedTocHref, TableOfContentsItem resolvedTopicItem, TableOfContentsModel subChildren) ProcessTocHref(SourceInfo<string> tocHref)
+            (SourceInfo<string> resolvedTocHref, TableOfContentsModel subChildren, (TableOfContentsItem item, Document doc) firstChild, TocHrefType tocHrefType) ProcessTocHref(SourceInfo<string> tocHref)
```
Comment (superyyrrzz) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: >(TableOfContentsItem item, Document doc)  [](start = 83, length = 41)

From the caller, seems it is not necessary to make return type a nested tuple.

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: used, return `default`

File: src/docfx/build/dependency/DependencyType.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/dependency/DependencyType.cs]: it's treated as inclusion in v2

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: it's necessary, firstItem is used separated

File: src/docfx/build/dependency/DependencyType.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/build/dependency/DependencyType.cs]: sorry, toc and toc relationships are missing in v2, so v3 generate it according to our own understanding.

**Comment by OsmondJiang** [comment]
@superyyrrzz v2 doesn't flatten toc inclusion dependencies, so this PR is focusing on flatten folder inclusion dependencies

**Comment by OsmondJiang** [comment]
the diff are expected:

image

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 5]
```diff
@@ -151,7 +169,25 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
+
+            return (errors, firstChild);
+
+            void ProcessSubChildren((TableOfContentsItem item, Document doc) subChildFromItems, (TableOfContentsItem item, Document doc) subChildFromTocHref)
+            {
+                if (subChildFromTocHref != default)
+                {
+                    subFirstChildren.Add(subChildFromTocHref);
+                }
+                else
```
Comment (live1206) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: else if (...)

**Pull request merged**